drivers/bicker_ser.c

```diff
@@ -0,0 +1,379 @@
+/*
+ * bicker_ser.c: support for Bicker SuperCapacitors DC UPSes
+ *
+ * Copyright (C) 2024 - Nicola Fontana <ntd@entidi.it>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+/* The protocol is detailed in section 14 of the DC/UPSIC user's manual,
+ * downloadable somewhere from the Bicker website.
+ *
+ * Basically, this is a binary protocol without checksums:
+ *
+ *  1 byte  1 byte  1 byte  1 byte  0..252 bytes  1 byte
+ * +-------+-------+-------+-------+--- - - - ---+-------+
+ * |  SOH  | Size  | Index |  CMD  |    Data     |  EOT  |
+ * +-------+-------+-------+-------+--- - - - ---+-------+
+ *
+ * where:
+ * - `SOH` is the start signal ('\x01')
+ * - `Size` is the length of the `Data` field (in bytes) plus 3
+ * - `Index` is the command index (always '\x03')
+ * - `CMD` is the command code to execute
+ * - `Data` is the (optional) argument of the command
+ * - `EOT` is the end signal ('\x04')
+ *
+ * The same format is used for incoming and outcoming packets. Although
+ * not explicitly documented, the data returned by the `Data` field
+ * seems to be in little-endian order.
+ */
+
+#include "config.h"
+#include "main.h"
+#include "attribute.h"
+
+#include "serial.h"
+
+#define DRIVER_NAME	"Bicker serial protocol"
+#define DRIVER_VERSION	"0.01"
+
+#define BICKER_PACKET	(1+1+1+1+252+1)
+#define BICKER_SOH	'\x01'
+#define BICKER_EOT	'\x04'
+#define BICKER_DELAY	20
+#define BICKER_RETRIES	3
+
+upsdrv_info_t upsdrv_info = {
+	DRIVER_NAME,
+	DRIVER_VERSION,
+	"Nicola Fontana <ntd@entidi.it>",
+	DRV_EXPERIMENTAL,
+	{ NULL }
+};
+
+static ssize_t bicker_send(char cmd, const void *data, size_t datalen)
+{
+	char buf[BICKER_PACKET];
+	size_t buflen;
+	ssize_t ret;
+
+	ser_flush_io(upsfd);
+
+	if (data != NULL) {
+		if (datalen > 252) {
```

Comment:
Would it make sense to define macros of `252` (max data length) and `3` below (header length?)

Reply:
It definitely does not hurt expressiveness: implemented in 651be816726ac9f77cc3d2837a77ec56e5022bfd.